modules/core/src/smithy4s/Validator.scala

```diff
@@ -0,0 +1,61 @@
+/*
+ *  Copyright 2021-2024 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s
+
+trait Validator[A, B] {
+  def validate(value: A): Either[String, B]
```

Comment:
**question:** given that a validator can potentially include multiple refinements, should we support `EitherNel` / `Either[::[String]` (due to the lack of a better NEL in the stdlib) from day 1, with error accumulation? A fail-fast alternative can also be given (with the current shape), for convenience.

Reply:
I thought about this, but decided against it initially due to a lack of a `NonEmptyList`. Forcing the caller to deal with a result of `Left(Nil)` felt kinda awkward, even though it would technically never happen.

Reply:
I agree with Denis. Generally speaking, and for better-or-worse, smithy4s does favour fail-fast as opposed to accumulation, so the decision is in line with the philosophy of the lib.

Reply:
it does... but I do recall us discussing the possibility of allowing accumulation :)

anyway, it's a sealed trait so we can extend it later.